### PR TITLE
hostname validation

### PR DESCRIPTION
**Pull request by wangqiang1208**
Changes in this PR:
1. installed a validates_hostname gem to validate hostname
2. specs

### CODE REVIEW

File: db/seeds.rb [hunk 1]
```diff
@@ -7,4 +7,4 @@
 #   Mayor.create(name: 'Emanuel', city: cities.first)
 #
 # Remember to ensure that the commands in this file are idempotent!
-Instance.find_or_create_by(name: 'Default', host: '*')
+Instance.find_or_create_by(name: 'Default', host: 'coursemology.org')
```
Comment (wangqiang1208) [inline db/seeds.rb]: @lowjoel  I think we need to validate the format of hostname ?   but here '*' is not a valid hostname, so I changed it.

Comment (lowjoel) [inline db/seeds.rb]: ...you just changed the meaning of my default instance.

what if a person sets up his own Coursemology instance, not at coursemology.org?

File: spec/models/instance_spec.rb [hunk 2]
```diff
@@ -1,10 +1,33 @@
...
+        expect(instance).to be_valid
+      end
+    end
+  end
+
   describe '.default' do
     it 'returns the default instance' do
       default_instance = Instance.default
-      expect(default_instance.host).to eq('*')
+      expect(default_instance.host).to eq('coursemology.org')
```
Comment (lowjoel) [inline spec/models/instance_spec.rb]: This is definitely factually wrong...

File: app/models/instance.rb [hunk 3]
```diff
@@ -15,6 +15,10 @@ def find_tenant_by_host(host)
     end
   end
 
+  VALID_HOST_NAME_REGEX = /\A[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.[a-z]{2,5}\z/i
+  validates :host, presence: true, format: { with: VALID_HOST_NAME_REGEX },
+    uniqueness: { case_sensitive: false }
```
Comment (lowjoel) [inline app/models/instance.rb]: Set a unique case sensitive index on the database table. no need to validate presence and uniqueness here, because the schema_validations gem does it.

the uniqueness validation is not thread safe anyway.

**Comment by lowjoel** [comment]
I'm not too sure you understand the rationale behind the default instance... it is for any instance, not just coursemology.org.

File: db/seeds.rb (hunk 1 above)
Comment (wangqiang1208) [inline db/seeds.rb]: So... what's the better way ? to change regular expression ?

Comment (lowjoel) [inline db/seeds.rb]: Add the validation so the users cannot make a wrong instance, but the seed 
needs to be there as part of the initial data set.

Comment (wangqiang1208) [inline db/seeds.rb]: I changed default hostname for a reason that it is not valid, so it cannot be saved to db.. then tests failed..

**Comment by wangqiang1208** [comment]
Hmm.. actually in my mind I think if user inputs an ip address

File: spec/models/instance_spec.rb [hunk 4]
```diff
@@ -1,10 +1,33 @@
...
+      hosts.each do |invalid_host|
+        instance.host = invalid_host
+        expect(instance).not_to be_valid
+      end
+    end
+  end
+
+  context 'when host format is valid' do
+    it 'is expected to be valid' do
+      hosts = %w[example.ORG www.example.org ex-ample.org example.org.sg]
```
Comment (lowjoel) [inline spec/models/instance_spec.rb]: I prefer using array of string literals actually

File: db/seeds.rb (hunk 1 above)
Comment (lowjoel) [inline db/seeds.rb]: nah, this has to be saved without validation, and cannot be changed in the UI.

**Comment by wangqiang1208** [comment]
I've updated

File: app/models/instance.rb [hunk 5]
```diff
@@ -15,6 +15,9 @@ def find_tenant_by_host(host)
     end
   end
 
+  VALID_HOST_NAME_REGEX = /\A[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.[a-z]{2,5}\z/i
```
Comment (lowjoel) [inline app/models/instance.rb]: fundamentally, this is what the RFC says, BUT actually it is possible to have an international domain name.

Comment (lowjoel) [inline app/models/instance.rb]: where is this regex from?

File: spec/models/instance_spec.rb [hunk 6]
```diff
@@ -1,6 +1,29 @@
...
+      hosts.each do |invalid_host|
+        instance.host = invalid_host
+        expect(instance).not_to be_valid
+      end
+    end
+  end
+
+  context 'when host format is valid' do
+    it 'is expected to be valid' do
+      hosts = ['example.ORG', 'www.example.org', 'ex-ample.org', 'example.org.sg']
```
Comment (lowjoel) [inline spec/models/instance_spec.rb]: I think the each should be outside the test. then for each you actually have one separate test case...

Comment (wangqiang1208) [inline spec/models/instance_spec.rb]: I'm not too agree with you here.. by doing this we are bring trouble to ourselves. 

Here if one of them failed the test failed, and it's enough for us to know there is a problem in validation.

File: app/models/instance.rb (hunk 5 above)
Comment (wangqiang1208) [inline app/models/instance.rb]: The regex is my modified version of one guys'  code in stackoverflow

Comment (lowjoel) [inline app/models/instance.rb]: no, please use the one specified in the RFC.

File: db/seeds.rb [hunk 7]
```diff
@@ -7,4 +7,6 @@
 #   Mayor.create(name: 'Emanuel', city: cities.first)
 #
 # Remember to ensure that the commands in this file are idempotent!
-Instance.find_or_create_by(name: 'Default', host: '*')
+
+# Default hostname without validation and cannot be changed in UI
+Instance.find_or_initialize_by(name: 'Default', host: '*').save(validate: false)
```
Comment (lowjoel) [inline db/seeds.rb]: `save!(validate: false)` so that if it fails the user will know.

File: spec/models/instance_spec.rb (hunk 6 above)
Comment (lowjoel) [inline spec/models/instance_spec.rb]: I think we want separate test cases for each of the failing hosts. This means it's all once test, when I think it's separate tests in the same context.

read it this way: you can actually say `it { should be_valid }` for each of those examples.

File: app/models/instance.rb (hunk 5 above)
Comment (wangqiang1208) [inline app/models/instance.rb]: OK, this one is strict than RFC

File: spec/models/instance_spec.rb (hunk 6 above)
Comment (wangqiang1208) [inline spec/models/instance_spec.rb]: Actually I tried, but that one is not possible, we cannot nested one test case inside a loop.

http://stackoverflow.com/questions/16698396/need-to-loop-through-an-array-in-rspec-test-not-running

File: app/models/instance.rb [hunk 8]
```diff
@@ -15,6 +15,9 @@ def find_tenant_by_host(host)
     end
   end
 
+  VALID_HOST_NAME_REGEX = /\A[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\z/i
+  validates :host, format: { with: VALID_HOST_NAME_REGEX }, length: { maximum: 255 }
```
Comment (wangqiang1208) [inline app/models/instance.rb]: See RFC 1123
http://tools.ietf.org/html/rfc1123#page-13

Comment (lowjoel) [inline app/models/instance.rb]: Why not just https://github.com/shuber/validates_as_hostname_label

Comment (weiqingtoh) [inline app/models/instance.rb]: Dissued this with @fonglh 

It will be a good idea to think about whether using Regex or Gem would be better. In both cases, the readability of the code is still maintained so no issues here. 

@wangqiang1208 's proposed solution is rather straightforward and elegant, but I think as we catch some edge cases this regex might expand.

The value of implementing the gem here would be that i) there is a full set of tests available, which the gem is assumed to have passed (so you do not need to re-think about the test cases), ii) the validations provide error messages (see [here](https://github.com/KimNorgaard/validates_hostname/blob/master/lib/validates_hostname.rb))

Given the time pressure, may I suggest that we use this first. This will save @wangqiang1208 some time in designing the error messages. At some point, if need be, we will dig into the code to fix other issues.

Comment (lowjoel) [inline app/models/instance.rb]: So much discussion.

### Why gems
My main driver for using gems are that:
 1. They provide an abstraction barrier. You use the gem, not needing to know how it is implemented. This reduces the cognitive load on the programmer
 2. This follows the single responsibility principle. The idea is, your code should handle one level of abstraction, no more, no less. Gems take the logic that you would otherwise have to implement in your application and factors it out into reusable components. In this case, models are closely associated with the database, but validation logic has little to do with validating what is a valid domain name.
 3. Helps cohesion, as a specific instance of point 2.

### Using many gems
That is true. But remember that gems are supposed to be functionally cohesive and should do one thing. Gems are meant to be reusable across projects and are meant to be general. If the gem does not solve the problem that you wished the gem solves, the standard procedure is to fork the gem and submit a patch. Rails was built up as a framework in this manner.

If you think about this further, the 'hard to understand the code' reason then seems shaky. If a programmer finds that the code is hard to understand because of the various abstractions that have been used in a large-scale program, then I think he needs to learn how to manage complexity better. Having to deal with so many things at once would be _worse_ for maintainability.

For example, in this case: if I'm looking at the Instance model, and I either:
 - see a call saying it's validated as a hostname. I would skip it and say, okay, assume that works. If everything else fails, I check the gem (how often does this happen...?)
 - Otherwise, I see that regex. I have to figure out what the regex means, and figure out why the regex is used. Then I add a comment saying what the regex does. Then the regex gets changed. Then the regex and the comment doesn't sync up... I leave it alone. Then the next programmer comes. Rinse and repeat.
   - Try doing this everywhere in your application!

Once again, I think we are considering too much about lowering the barrier to entry for contributors. That is useful, and I've made such a decisions in the past; but history and experience shows that making things accessible brings absolutely no new contributors. If your have a complex project that solves problems, people will come. Perhaps that is true from a systems programming perspective (definitely true when I was writing kernel modules and other components in C++) but when you have a large scale application, this reason would prevent the application from doing useful things. _Complexity is inevitable, managing it through abstractions is how you ensure you can deal with it!_

### What to consider when using gems

 1. Using a little of the gem should not preclude the use of a gem. How much of Rails do you think we will be using? :smile: 
 2. How much time we would need to write the code out would probably be a concern. But remember that by doing this you are merely trading short-term speed for long-term maintenance. While this may be the agile way (don't write what you don't need), I think that for most instances, the barrier to using a gem is actually quite low, and you'll be spending about the same amount of time anyway.
 3. Future development is important.
 4. Maintenance is important. But sometimes a gem is just too good not to use. In the worst case, fork and _be the maintainer_. This makes you a better programmer and software engineer as well. You make decisions that typically you might expect a more senior programmer to make. You'll make mistakes. You'll learn. And then you won't. And then _you'll_ be the senior programmer.

Ultimately, whether to use a gem is a subjective opinion, and depends on the experiences of the people who have used others. Having used good and bad gems helps in making that decision. It's not necessarily something that can be done via consensus, unfortunately.

Comment (wangqiang1208) [inline app/models/instance.rb]: @lowjoel Thanks for the detailed comments also !

So you are encouraging us to use gems as many as we can  ?

>Otherwise, I see that regex. I have to figure out what the regex means, and figure out why the regex is used. Then I add a comment saying what the regex does. Then the regex gets changed. Then the regex and the comment doesn't sync up... I leave it alone. Then the next programmer comes. Rinse and repeat.

BTW, I'm actually not too agree with this,  we can also achieve the same goal by proper name and document the function, you do not need to look at the details of regex and can just assume that the function works. This requires more application code, but makes debug easier.

> see a call saying it's validated as a hostname. I would skip it and say, okay, assume that works. If everything else fails, I check the gem (how often does this happen...?)

And here I think If we are using functions from gems, we should provide some document? 
Because to me most of time I don't know where is the function from and what it does.. normally I need to search the function name online and then hopefully I'll find the gem's document, but it's painful...

Comment (lowjoel) [inline app/models/instance.rb]: I'm not saying use as many gems as possible. Evaluate them. Sometimes it's trivial, and should not require a dependency. Others are not so. This case it might look trivial but you are implementing an RFC -- not something I'll try to implement in one regex.

Yes, proper naming is a basic requirement. But the problem is that by implementing this separately the Regex is going to become eventually more and more complicated as we add more and more support. At this point, I already cannot completely understand what the regex does, or why it is correct. I know what it's _for_, but not _why_ it works. To be exposed to such a detail while reading the Instance class makes me feel like I'm being presented with too much information. In a separate component, you can break the regex down to different parts, for example. Not something you can do at this level of abstraction.

I don't know about you, but I think the first thing I'd do to find what function does what is to read the Gemfile... that's not saved me all the time, and I still do searches. But what documentation do you propose? Oh, and a debugger helps too. If gems are properly used, the contributor will see the same functions repeat again and again... and the learning cost will be part of the onboarding process. I don't think "more documentation" is the solution to spreading knowledge... because honestly who _wants_ to read documentation, and worse, who _wants_ to write documentation? Do you read documentation when opening a project once? Or do you only start reading when you can't deal with something?

Comment (wangqiang1208) [inline app/models/instance.rb]: Yes, I start reading documentation when I cannot deal with it :-)

So I think document where is this function from and basically what is does would be enough ? Should not require any more effort to maintain.

Comment (lowjoel) [inline app/models/instance.rb]: that naturally begs the question: where?

if you put it in your code, then you'll have the same comment appear in every single file... which is a lot of clutter.

Comment (wangqiang1208) [inline app/models/instance.rb]: For gems across components, that one hard to doc, or maybe we doc something
in every module's description ?

For gems only for one feature, like our case here, I think we can directly
put comments before the function from gem.

Comment (lowjoel) [inline app/models/instance.rb]: modules are split up, like models/controllers/views.

actually, i still think the simplest is just trace into the function call in a debugger lah. it's not hard to put breakpoint and step-into...

Comment (jsyeo) [inline app/models/instance.rb]: Or you can use Ruby mine to see where it is defined.

Comment (lowjoel) [inline app/models/instance.rb]: not always possible, have you seen `alias_method_chain`? or `instance_eval` or `class_eval` used with strings (yes...)

ah, the powers of Ruby metaprogramming.

Comment (wangqiang1208) [inline app/models/instance.rb]: OK lah, let's pray for the new guys :-)

Comment (wangqiang1208) [inline app/models/instance.rb]: Unfortunately, gem 'validates_hostname' has a bug. it considers host name ends with dots to be valid >_<

For example:  'example.org....' is valid

I've submitted an issue in their repo

Comment (lowjoel) [inline app/models/instance.rb]: Good.  A bigger learning experience would be how to fix it =p

Comment (wangqiang1208) [inline app/models/instance.rb]: Hmm, I'll fix it in my own repo first.

Comment (lowjoel) [inline app/models/instance.rb]: Actually I remember something. Trailing period is acceptable. It indicates a fully fully qualified name without the connections DNS suffix. You use this in raw DNS records.

Comment (wangqiang1208) [inline app/models/instance.rb]: > A "name" (Net, Host, Gateway, or Domain name) is a text string up
   to 24 characters drawn from the alphabet (A-Z), digits (0-9), minus
   sign (-), and period (.).  Note that periods are only allowed when
   they serve to delimit components of "domain style names". (See
   RFC-921)

See RFC document

Comment (lowjoel) [inline app/models/instance.rb]: ... And you are using a domain style name, and it is serving as a delimiter with an empty last element...

Comment (wangqiang1208) [inline app/models/instance.rb]: The content between dots are called labels.. I don't think it allows empty
labels?

Comment (lowjoel) [inline app/models/instance.rb]: What label do you call the root? I don't know. I'm guessing. But trailing periods make sense to me, I've seen them in use before so I've no reason to believe they're invalid.

What you're looking for is probably normalisation, not validation.

Comment (wangqiang1208) [inline app/models/instance.rb]: Can you provide any links ?

I still think trailing periods is invalid, and I also tried some validation
tools online to confirm.

Comment (lowjoel) [inline app/models/instance.rb]: http://en.wikipedia.org/wiki/Fully_qualified_domain_name my guess is correct.

Comment (wangqiang1208) [inline app/models/instance.rb]: Alright, such domain name exists.

As it says, it was used for DNS resolution, so I guess there is no website
in real life whose domain name has a trailing dot ?

Comment (lowjoel) [inline app/models/instance.rb]: All the websites have a trailing dot. The process of normalisation adds it there.

 

As mentioned earlier, I think there’s some confusion between validation and normalisation. The FQDN has a trailing dot. The domain names we use do not because it is normalised to one with a trailing dot.

File: Gemfile.lock [hunk 9]
```diff
@@ -244,6 +244,9 @@ GEM
       actionpack (>= 3.0)
       activesupport (>= 3.0)
       sprockets (>= 2.8, < 4.0)
+    sqlite3 (1.3.10)
```
Comment (lowjoel) [inline Gemfile.lock]: ...why sqlite3?

Comment (wangqiang1208) [inline Gemfile.lock]: It comes with the gem...

Comment (lowjoel) [inline Gemfile.lock]: I give you a small challenge: try to fix the gemspec such that it doesn’t need it, then PR.

Hint: it’s a development dependency (I think…)

Comment (wangqiang1208) [inline Gemfile.lock]: Sure, let me see.

Comment (lowjoel) [inline Gemfile.lock]: oh, it's already fixed. Use the gem from github.

File: Gemfile [hunk 10]
```diff
@@ -96,6 +96,9 @@ end
 # Multitenancy
 gem 'acts_as_tenant'
 
+# Extension for validating hostnames and domain names
+gem 'validates_hostname', '~> 1.0.0'
```
Comment (lowjoel) [inline Gemfile]: I personally rather not use the version specifier unless necessary. So I'll just put `gem 'validates_hostname'`. If it breaks because of an update, we should update our code.

Rails is probably the only exception.

**Comment by lowjoel** [comment]
also you didn't specify in the databse that the hostname must be unique and case insensitive.

File: Gemfile.lock (hunk 9 above)
Comment (lowjoel) [inline Gemfile.lock]: and there's a bug in there preventing us from using in Rails 4. PR submitted... KimNorgaard/validates_hostname#4

File: spec/models/instance_spec.rb [hunk 11]
```diff
@@ -1,6 +1,38 @@
 require 'rails_helper'
 
 RSpec.describe Instance, type: :model do
+  describe 'hostname validation' do
+    let(:instance) { build(:instance) }
+
+    context 'when hostname format is invalid' do
```
Comment (lowjoel) [inline spec/models/instance_spec.rb]:
```ruby
    context 'when hostname format is invalid' do
      ['example,com', 'www_.example.org', 'example_.org',
        'example.org_', 'example .org', 'example. org'].each do |domain|
        context domain do
          let(:instance) do
            instance = build(:instance)
            instance.host = domain
            instance
          end
          subject { instance }
          it { should_not be_valid }
        end
      end
    end
```

Comment (lowjoel) [inline spec/models/instance_spec.rb]: gives you information about which failed.

**Comment by wangqiang1208** [comment]
Redirected the gem to your repo.
The host name is already unique and case insensitive in database, We've done that before.

**Comment by lowjoel** [comment]
The gem has ben updated. Can use the upstream version. See if there’s a new version, if there is then use that. Otherwise use github repo.

**Comment by wangqiang1208** [comment]
Yep, I saw your PR has been merged.

**Comment by wangqiang1208** [comment]
Nope, there is no new version of that gem. I'll use github version then

**Comment by wangqiang1208** [comment]
The gem has been updated, he fixed the issue and replied me this morning.

**Comment by lowjoel** [comment]
i think you need to rebase.

**Comment by wangqiang1208** [comment]
Oh sorry, didn't notice the conflicts.
